### PR TITLE
Unit test for progressRedeemWithProof

### PR DESCRIPTION
**Pull request by deepesh-kn**
⚠️ Blocked by #578

This PR includes unit test case for  `EIP20CoGateway progressRedeemWithProof()`
Fixes: #438

### CODE REVIEW

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 1]
```diff
@@ -0,0 +1,541 @@
+// Copyright 2018 OpenST Ltd.
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]:
```suggestion
// Copyright 2019 OpenST Ltd.
```

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]:
```suggestion
```

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+      hashLock: proofData.co_gateway.redeem.params.hashLock,
+      messageHash: proofData.co_gateway.progress_redeem.params.messageHash,
+      rlpParentNodes: proofData.gateway.confirm_redeem_intent.proof_data.storageProof[0].serializedProof,
+      blockHeight: new BN(proofData.gateway.confirm_redeem_intent.proof_data.block_number, 16),
+      messageStatus: MessageStatusEnum.Declared
+    };
+
+    utilityToken = await MockUtilityToken.new(
+      accounts[9],
+      "",
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: Please add dummy value, named parameter/comments.

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+    );
+
+    assert.equal(
+      tx.receipt.status,
+      1,
+      "Receipt status is unsuccessful",
+    );
+
+  });
+
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]:
```suggestion
```

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+    );
+
+    assert.strictEqual(
+      eventData._redeemer,
+      redeemParams.redeemer,
+      `Redeemer address ${eventData._redeemer} from event is not equal to expected redeemer address ${redeemParams.redeemer}`,
+    );
+
+    assert.strictEqual(
+      eventData._redeemerNonce.eq(redeemParams.nonce),
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: Please add `toString(10)` for BN in the reason.

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+      blockHeight: new BN(proofData.gateway.confirm_redeem_intent.proof_data.block_number, 16),
+      messageStatus: MessageStatusEnum.Declared
+    };
+
+    utilityToken = await MockUtilityToken.new(
+      accounts[9], // Token address.
+      "DMY", // Symbol.
+      "Dummy token", // Token name.
+      18, // Token decimal.
+      accounts[2], // Organisation address.
```
Comment (schemar) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: An alternative to the comments would be named variables.

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+      MessageStatusEnum.Declared,
+    );
+
+    // Set co-gateway to owner so that increase supply can be called.
+    await utilityToken.setCoGatewayAddress(owner);
+    // Send redeem amount to co-gateway.
+    await  utilityToken.increaseSupply(eip20CoGateway.address, redeemParams.amount, {from: owner});
+    // Send bounty to co-gateway.
+    await web3.eth.sendTransaction(
+      {to: eip20CoGateway.address, from: facilitator, value: bountyAmount}
```
Comment (schemar) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: Also more readable when split.
```suggestion
      {
        to: eip20CoGateway.address,
        from: facilitator,
        value: bountyAmount,
      },
```

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+    );
+    await utilityToken.setCoGatewayAddress(eip20CoGateway.address);
+
+  });
+
+  it('should fail when message hash is zero ', async function () {
```
Comment (schemar) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: Why extra spaces at the ends of all `it` descriptions?
```suggestion
  it('should fail when message hash is zero', async function () {
```

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+    assert.strictEqual(
+      result.redeemer_,
+      redeemParams.redeemer,
+      `Redeemer address must be equal to ${redeemParams.redeemer}`,
+    );
+
+    assert.strictEqual(
+      result.redeemAmount_.eq(redeemParams.amount),
+      true,
+      `Redeem amount must be equal to ${redeemParams.amount.toString(10)}`,
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: Please add both values of assert.

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: Same comment

File: test/gateway/eip20_cogateway/progress_redeem_with_proof.js [hunk 11]
```diff
@@ -0,0 +1,556 @@
...
+      new BN(proofData.gateway.progress_unstake.proof_data.block_number, 16),
+      MessageStatusEnum.Progressed,
+      { from: facilitator },
+    );
+
+    let finalTotalSupply = await utilityToken.totalSupply.call();
+    let finalCoGatewayBalance = await utilityToken.balanceOf(eip20CoGateway.address);
+
+    assert.strictEqual(
+      finalTotalSupply.eq(initialTotalSupply.sub(redeemParams.amount)),
```
Comment (sarvesh-ost) [inline test/gateway/eip20_cogateway/progress_redeem_with_proof.js]: Similar comment.

**Pull request merged**